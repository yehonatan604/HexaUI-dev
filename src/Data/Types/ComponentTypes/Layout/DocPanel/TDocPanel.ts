import { TComponent, ComponentOptions } from "../../TComponent";

export type TDocPanel = TComponent<"div"> & {
    options: DocPanelOptions;
    panelWidth?: string;
    innerClassName?: string;
}

export type DocPanelOptions = ComponentOptions & {
    posiotion?: 'relative' | 'absolute' | 'fixed' | 'sticky';
    placement: 'top' | 'right' | 'bottom' | 'left';
}