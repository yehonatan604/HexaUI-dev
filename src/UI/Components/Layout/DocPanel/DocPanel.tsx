import React, { useEffect, useRef, useState, useCallback } from "react";
import { FlexDir } from "../../../../Data/Constants/FlexDirection";
import { TDocPanel } from "./types/TDocPanel";
import Flex from "../Flex/Flex";

const DocPanel = ({
  options,
  className = "",
  innerClassName = "",
  panelWidth: initialWidth,
  children,
  ...componentProps
}: TDocPanel) => {
  const [panelWidth, setPanelWidth] = useState(initialWidth || "20%");
  const panelRef = useRef<HTMLDivElement>(null);
  const borderRef = useRef<HTMLDivElement>(null);
  const isDraggingRef = useRef(false);

  const direction = (() => {
    switch (options.placement) {
      case "top":
        return FlexDir.Col;
      case "bottom":
        return FlexDir.ColReverse;
      case "left":
        return FlexDir.Row;
      default:
        return FlexDir.RowReverse;
    }
  })();

  const isHorizontal = direction === FlexDir.Row || direction === FlexDir.RowReverse;

  const handleResize = useCallback(
    (e: MouseEvent) => {
      if (!isDraggingRef.current || !panelRef.current) return;

      const rect = panelRef.current.getBoundingClientRect();
      let newSize;

      if (isHorizontal) {
        newSize =
          direction === FlexDir.Row ? e.clientX - rect.left : rect.right - e.clientX;
      } else {
        newSize =
          direction === FlexDir.Col ? e.clientY - rect.top : rect.bottom - e.clientY;
      }

      // Ensure the new size is within reasonable bounds
      newSize = Math.max(
        100,
        Math.min(
          newSize,
          isHorizontal ? window.innerWidth - 100 : window.innerHeight - 100
        )
      );

      if (isHorizontal) setPanelWidth(`${newSize}px`);
    },
    [direction, isHorizontal]
  );

  const handleMouseUp = useCallback(() => {
    isDraggingRef.current = false;
    document.removeEventListener("mousemove", handleResize);
    document.removeEventListener("mouseup", handleMouseUp);
    document.body.style.userSelect = "";
    document.body.style.cursor = "";
  }, [handleResize]);

  const handleMouseDown = useCallback(() => {
    isDraggingRef.current = true;
    document.addEventListener("mousemove", handleResize);
    document.addEventListener("mouseup", handleMouseUp);
    document.body.style.userSelect = "none";
    document.body.style.cursor = isHorizontal ? "ew-resize" : "ns-resize";
  }, [handleResize, handleMouseUp, isHorizontal]);

  useEffect(() => {
    const border = borderRef.current;
    if (border) {
      border.addEventListener("mousedown", handleMouseDown);
    }
    return () => {
      if (border) {
        border.removeEventListener("mousedown", handleMouseDown);
      }
    };
  }, [handleMouseDown]);

  return (
    <Flex
      options={{
        direction: direction,
        align: "t",
      }}
      className={`relative ${className} w-full`}
    >
      <div
        ref={panelRef}
        className={`flex flex-col bg-${
          options.bgVariant || "black-d"
        } ${innerClassName} text-nowrap text-ellipsis overflow-hidden`}
        {...componentProps}
        style={{
          [isHorizontal ? "width" : "height"]: panelWidth,
        }}
      >
        {children}
      </div>
      <div
        ref={borderRef}
        className={`bg-${options.border?.variant || "black-d"} ${
          isHorizontal
            ? `w-${options.border?.thickness || 1} cursor-ew-resize`
            : `h-${options.border?.thickness || 1} cursor-ns-resize`
        }`}
        style={{ flexShrink: 0 }}
      />
    </Flex>
  );
};

export default DocPanel;
