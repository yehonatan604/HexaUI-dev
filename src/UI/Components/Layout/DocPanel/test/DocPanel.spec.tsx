import { render, screen, fireEvent } from "@testing-library/react";
import DocPanel from "../DocPanel"; // Adjust the import path
import { describe, it, expect } from "vitest";

describe("DocPanel Component", () => {
  it("should render the panel with the correct placement", () => {
    render(<DocPanel options={{ placement: "right" }}>Panel Content</DocPanel>);

    // Check if the panel content is rendered
    expect(screen.getByText("Panel Content")).toBeInTheDocument();

    // Check if the panel has the correct class for background
    const panelElement = screen.getByText("Panel Content").parentElement;
    expect(panelElement).toBeInTheDocument();
  });

  it("should render with the correct initial width", () => {
    render(
      <DocPanel options={{ placement: "right" }} panelWidth="300">
        Panel Content
      </DocPanel>
    );

    const panelElement = screen.getByText("Panel Content").parentElement;

    // Check if the initial width is set correctly
    expect(panelElement).toHaveStyle({ width: "300" });
  });

  it("should resize the panel when dragged", () => {
    render(
      <DocPanel options={{ placement: "right" }} panelWidth="300">
        Panel Content
      </DocPanel>
    );

    const panelElement = screen.getByText("Panel Content").parentElement;
    const borderElement = panelElement?.lastChild; // Grabbing the resizable border

    // Check if the initial width is 300px
    expect(panelElement).toHaveStyle({ width: "300" });

    // Ensure that the border element exists
    expect(borderElement).toBeInTheDocument();

    // Simulate the mousedown event on the border to start dragging
    fireEvent.mouseDown(borderElement!);

    // Simulate a mousemove event to resize the panel
    fireEvent.mouseMove(document, { clientX: 400 });

    // Check if the width has been updated correctly after dragging
    expect(panelElement).toHaveStyle({ width: "400" });

    // Simulate mouseup event to end dragging
    fireEvent.mouseUp(document);
  });

  it("should render and match the snapshot", () => {
    const { asFragment } = render(
      <DocPanel options={{ placement: "right" }} panelWidth="300">
        Panel Content
      </DocPanel>
    );

    // Snapshot the DocPanel structure
    expect(asFragment()).toMatchSnapshot();
  });
});
