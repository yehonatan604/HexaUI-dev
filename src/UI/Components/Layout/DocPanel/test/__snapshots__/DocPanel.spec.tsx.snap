// Vitest Snapshot v1

exports[`DocPanel Component > should render and match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row-reverse bg- text- justify-start items-t flex-nowrap relative  w-full"
  >
    <div
      class="flex flex-col bg-black-d  text-nowrap text-ellipsis overflow-hidden"
    >
      Panel Content
    </div>
    <div
      class="bg-black-d w-1 cursor-ew-resize"
      style="flex-shrink: 0;"
    />
  </div>
</DocumentFragment>
`;
